BUG: `convert_dtypes()` converts GeoDataFrame to DataFrame

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the latest version of geopandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of geopandas.

---

#### Code Sample, a copy-pastable example

```python
import pandas as pd
import fiona # AttributeError if not importing fiona before gpd
import geopandas as gpd
from geopandas.testing import assert_geodataframe_equal

def df():
    return pd.DataFrame(
        {
            "City": ["Buenos Aires", "Brasilia", "Santiago", "Bogota", "Caracas"],
            "Country": ["Argentina", "Brazil", "Chile", "Colombia", "Venezuela"],
            "Latitude": [-34.58, -15.78, -33.45, 4.60, 10.48],
            "Longitude": [-58.66, -47.91, -70.66, -74.08, -66.86],
        }
    )

def gdf_points_from_xy(df):
    return gpd.GeoDataFrame(df, geometry=gpd.points_from_xy(df.Longitude, df.Latitude))

def test_convert_dtypes_before_gdf():
    result = df().convert_dtypes().pipe(gdf_points_from_xy)
    assert isinstance(result, gpd.GeoDataFrame)
    # --> no error

def test_convert_dtypes_after_gdf():
    result = df().pipe(gdf_points_from_xy).convert_dtypes()
    assert isinstance(result, gpd.GeoDataFrame)
    # --> AssertionError

def test_convert_dtypes_expectation():
    expected = df().convert_dtypes().pipe(gdf_points_from_xy)

    result = df().pipe(gdf_points_from_xy).convert_dtypes()

    assert_geodataframe_equal(result, expected)
    #--> AssertionError
```

#### Problem description

Calling `.convert_dtypes()` on a GeoDataFrame turns it into a regular DataFrame.

```python
AssertionError: assert isinstance(result, GeoDataFrame)
```

#### Expected Output
I expect `convert_dtypes` to not change frame type.

The problem may be with the [original method](https://github.com/pandas-dev/pandas/blob/c5d0d138cf1162cea985741cbbbaea3eef6ca8c3/pandas/core/generic.py#L6138). I suspect it is because of `return concat(results, axis=1, copy=False)`. As the pandas repo does not know about other frame types, I suspect the fix should lie in this repo.

#### Output of ``geopandas.show_versions()``

<details>

SYSTEM INFO
-----------
python     : 3.8.4 (default, Jan 11 2021, 16:58:12)  [Clang 12.0.0 (clang-1200.0.32.28)]
executable : /Users/adriantofting/Library/Caches/pypoetry/virtualenvs/gpd-test-qc5veAeh-py3.8/bin/python
machine    : macOS-11.1-x86_64-i386-64bit

GEOS, GDAL, PROJ INFO
---------------------
GEOS       : 3.9.1
GEOS lib   : /usr/local/Cellar/geos/3.9.1/lib/libgeos_c.dylib
GDAL       : 3.2.1
GDAL data dir: None
PROJ       : 7.2.1
PROJ data dir: /usr/local/share/proj

PYTHON DEPENDENCIES
-------------------
geopandas  : 0.9.0
pandas     : 1.2.3
fiona      : 1.8.18
numpy      : 1.20.1
shapely    : 1.7.1
rtree      : None
pyproj     : 3.0.0.post1
matplotlib : None
mapclassify: None
geopy      : None
psycopg2   : None
geoalchemy2: None
pyarrow    : None
pygeos     : None

</details>

#### Environment (poetry)

<details>

```toml
#pyproject.toml
[tool.poetry]
name = "gpd_test"
version = "0.1.0"
description = ""

[tool.poetry.dependencies]
python = "^3.8"
geopandas = "0.9.0"

[tool.poetry.dev-dependencies]
jupyterlab = "^3.0.9"

[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"
```

</details

--- a/geopandas/geodataframe.py
+++ b/geopandas/geodataframe.py
@@ -1663,7 +1663,7 @@ box': (2.0, 1.0, 2.0, 1.0)}], 'bbox': (1.0, 1.0, 2.0, 2.0)}
         """
         Convert columns to best possible dtypes using dtypes supporting ``pd.NA``.
 
-        Returns a GeoDataFrame always, as no conversions are applied to the
+        Always returns a GeoDataFrame as no conversions are applied to the
         geometry column.
 
         See the pandas.DataFrame.convert_dtypes docstring for more details.

--- a/geopandas/geodataframe.py
+++ b/geopandas/geodataframe.py
@@ -1652,14 +1652,7 @@ box': (2.0, 1.0, 2.0, 1.0)}], 'bbox': (1.0, 1.0, 2.0, 2.0)}
         # do not return a GeoDataFrame
         return pd.DataFrame(df)
 
-    def convert_dtypes(
-        self,
-        infer_objects=True,
-        convert_string=True,
-        convert_integer=True,
-        convert_boolean=True,
-        convert_floating=True,
-    ):
+    def convert_dtypes(self, *args, **kwargs):
         """
         Convert columns to best possible dtypes using dtypes supporting ``pd.NA``.
 
@@ -1676,29 +1669,13 @@ box': (2.0, 1.0, 2.0, 1.0)}], 'bbox': (1.0, 1.0, 2.0, 2.0)}
         # Overridden to fix GH1870, that return type is not preserved always
         # (and where it was, geometry col was not)
 
-        kwargs = dict(
-            infer_objects=infer_objects,
-            convert_string=convert_string,
-            convert_integer=convert_integer,
-            convert_boolean=convert_boolean,
-            convert_floating=convert_floating,
-        )
-        print(compat.PANDAS_GE_10)
         if not compat.PANDAS_GE_10:
             raise NotImplementedError(
                 "GeoDataFrame.convert_dtypes requires pandas >= 1.0"
             )
-        if not compat.PANDAS_GE_11:
-            kwargs.pop("convert_floating")
-            if (
-                not convert_floating
-            ):  # If they passed a non-default convert_floating value
-                warnings.warn(
-                    "convert_floating kwarg is not supported for pandas < 1.1.0"
-                )
 
         return GeoDataFrame(
-            super().convert_dtypes(kwargs),
+            super().convert_dtypes(*args, **kwargs),
             geometry=self.geometry.name,
             crs=self.crs,
         )

--- a/geopandas/geodataframe.py
+++ b/geopandas/geodataframe.py
@@ -1673,16 +1673,32 @@ box': (2.0, 1.0, 2.0, 1.0)}], 'bbox': (1.0, 1.0, 2.0, 2.0)}
         GeoDataFrame
 
         """
-        # Overidden to fix GH1870, that return type is not preserved always
+        # Overridden to fix GH1870, that return type is not preserved always
         # (and where it was, geometry col was not)
+
+        kwargs = dict(
+            infer_objects=infer_objects,
+            convert_string=convert_string,
+            convert_integer=convert_integer,
+            convert_boolean=convert_boolean,
+            convert_floating=convert_floating,
+        )
+        print(compat.PANDAS_GE_10)
+        if not compat.PANDAS_GE_10:
+            raise NotImplementedError(
+                "GeoDataFrame.convert_dtypes requires pandas >= 1.0"
+            )
+        if not compat.PANDAS_GE_11:
+            kwargs.pop("convert_floating")
+            if (
+                not convert_floating
+            ):  # If they passed a non-default convert_floating value
+                warnings.warn(
+                    "convert_floating kwarg is not supported for pandas < 1.1.0"
+                )
+
         return GeoDataFrame(
-            super().convert_dtypes(
-                infer_objects,
-                convert_string,
-                convert_integer,
-                convert_boolean,
-                convert_floating,
-            ),
+            super().convert_dtypes(kwargs),
             geometry=self.geometry.name,
             crs=self.crs,
         )

--- a/geopandas/geodataframe.py
+++ b/geopandas/geodataframe.py
@@ -1652,6 +1652,41 @@ box': (2.0, 1.0, 2.0, 1.0)}], 'bbox': (1.0, 1.0, 2.0, 2.0)}
         # do not return a GeoDataFrame
         return pd.DataFrame(df)
 
+    def convert_dtypes(
+        self,
+        infer_objects=True,
+        convert_string=True,
+        convert_integer=True,
+        convert_boolean=True,
+        convert_floating=True,
+    ):
+        """
+        Convert columns to best possible dtypes using dtypes supporting ``pd.NA``.
+
+        Returns a GeoDataFrame always, as no conversions are applied to the
+        geometry column.
+
+        See the pandas.DataFrame.convert_dtypes docstring for more details.
+
+        Returns
+        -------
+        GeoDataFrame
+
+        """
+        # Overidden to fix GH1870, that return type is not preserved always
+        # (and where it was, geometry col was not)
+        return GeoDataFrame(
+            super().convert_dtypes(
+                infer_objects,
+                convert_string,
+                convert_integer,
+                convert_boolean,
+                convert_floating,
+            ),
+            geometry=self.geometry.name,
+            crs=self.crs,
+        )
+
     def to_postgis(
         self,
         name,

--- a/geopandas/tests/test_pandas_methods.py
+++ b/geopandas/tests/test_pandas_methods.py
@@ -267,23 +267,29 @@ def test_convert_dtypes(df):
 
     # Test geom first, geom_col=geometry (order is important in concat,
     #   used internally)
+    expected1 = GeoDataFrame(
+        pd.DataFrame(df).convert_dtypes(), crs=df.crs, geometry=df.geometry.name
+    )
     res1 = df.convert_dtypes()
     # Checking type and metadata are right
-    assert_geodataframe_equal(df, res1)
+    assert_geodataframe_equal(expected1, res1)
 
     # Test geom last, geom_col=geometry
     res2 = df[["value1", "value2", "geometry"]].convert_dtypes()
-    assert_geodataframe_equal(df[["value1", "value2", "geometry"]], res2)
+    assert_geodataframe_equal(expected1[["value1", "value2", "geometry"]], res2)
 
     # Test again with crs set and custom geom col name
     df2 = df.set_crs(epsg=4326).rename_geometry("points")
+    expected2 = GeoDataFrame(
+        pd.DataFrame(df2).convert_dtypes(), crs=df2.crs, geometry=df2.geometry.name
+    )
     res3 = df2.convert_dtypes()
     # Checking type and metadata are right
-    assert_geodataframe_equal(df2, res3)
+    assert_geodataframe_equal(expected2, res3)
 
     # Test geom last, geom_col=geometry
-    res4 = df2[["value1", "value2", "geometry"]].convert_dtypes()
-    assert_geodataframe_equal(df2[["value1", "value2", "geometry"]], res4)
+    res4 = df2[["value1", "value2", "points"]].convert_dtypes()
+    assert_geodataframe_equal(expected2[["value1", "value2", "points"]], res4)
 
 
 def test_to_csv(df):

--- a/geopandas/tests/test_pandas_methods.py
+++ b/geopandas/tests/test_pandas_methods.py
@@ -262,19 +262,26 @@ def test_astype_invalid_geodataframe():
     assert res["a"].dtype == object
 
 
+@pytest.mark.xfail(
+    not compat.PANDAS_GE_10,
+    reason="Convert dtypes new in pandas 1.0",
+    raises=NotImplementedError,
+)
 def test_convert_dtypes(df):
     # https://github.com/geopandas/geopandas/issues/1870
 
-    # Test geom first, geom_col=geometry (order is important in concat,
-    #   used internally)
+    # Test geometry col is first col, first, geom_col_name=geometry
+    # (order is important in concat, used internally)
+    res1 = df.convert_dtypes()  # note res1 done first for pandas < 1 xfail check
+
     expected1 = GeoDataFrame(
         pd.DataFrame(df).convert_dtypes(), crs=df.crs, geometry=df.geometry.name
     )
-    res1 = df.convert_dtypes()
+
     # Checking type and metadata are right
     assert_geodataframe_equal(expected1, res1)
 
-    # Test geom last, geom_col=geometry
+    # Test geom last, geom_col_name=geometry
     res2 = df[["value1", "value2", "geometry"]].convert_dtypes()
     assert_geodataframe_equal(expected1[["value1", "value2", "geometry"]], res2)
 
@@ -284,7 +291,6 @@ def test_convert_dtypes(df):
         pd.DataFrame(df2).convert_dtypes(), crs=df2.crs, geometry=df2.geometry.name
     )
     res3 = df2.convert_dtypes()
-    # Checking type and metadata are right
     assert_geodataframe_equal(expected2, res3)
 
     # Test geom last, geom_col=geometry

--- a/geopandas/tests/test_pandas_methods.py
+++ b/geopandas/tests/test_pandas_methods.py
@@ -262,6 +262,30 @@ def test_astype_invalid_geodataframe():
     assert res["a"].dtype == object
 
 
+def test_convert_dtypes(df):
+    # https://github.com/geopandas/geopandas/issues/1870
+
+    # Test geom first, geom_col=geometry (order is important in concat,
+    #   used internally)
+    res1 = df.convert_dtypes()
+    # Checking type and metadata are right
+    assert_geodataframe_equal(df, res1)
+
+    # Test geom last, geom_col=geometry
+    res2 = df[["value1", "value2", "geometry"]].convert_dtypes()
+    assert_geodataframe_equal(df[["value1", "value2", "geometry"]], res2)
+
+    # Test again with crs set and custom geom col name
+    df2 = df.set_crs(epsg=4326).rename_geometry("points")
+    res3 = df2.convert_dtypes()
+    # Checking type and metadata are right
+    assert_geodataframe_equal(df2, res3)
+
+    # Test geom last, geom_col=geometry
+    res4 = df2[["value1", "value2", "geometry"]].convert_dtypes()
+    assert_geodataframe_equal(df2[["value1", "value2", "geometry"]], res4)
+
+
 def test_to_csv(df):
 
     exp = (

